### ISSUE TITLE
Poor Consumer Test Performance on Windows

### ISSUE THREAD

**Issue by adamrodger** [pact-foundation/pact-net#323]
Consumer tests with the new Rust FFI have very poor performance when calling the mock server via `HttpClient` on Windows. This is quite a large regression on the Ruby backend performance, and one of the key benefits of the Rust FFI backend is supposed to be increased performance.

## Demonstration of performance

It typically takes around 2 seconds for the `HttpResponse` to come back, which means every consumer test takes a minimum of 2 seconds. It's not uncommon to have dozens of consumer tests per solution, and testing libraries typically won't run tests within the same file/collection in parallel, and so it would result in the consumer test run taking minutes to complete.

Here's an example:

![image](https://user-images.githubusercontent.com/508586/133938806-591a2359-3362-442a-ae2a-ee08d4c20c39.png)

And with some logging added to show where the problem happen:

![image](https://user-images.githubusercontent.com/508586/133938993-d8a1f902-055c-470f-bd22-a1dc29c74415.png)

![image](https://user-images.githubusercontent.com/508586/133938983-410d4704-dd1a-45dd-82e0-5e2ad682c6a6.png)

## Only on Windows

It's worth noting that this only seems to happen on Windows. On Linux the consumer tests run extremely quickly. Check [recent CI runs](https://github.com/pact-foundation/pact-net/actions) to confirm, but typically the Ubuntu run finishes in a fraction of the time of the Windows run. I have confirmed locally that the tests are extremely slow on my Windows machine as well, so it's not some issue with GitHub Actions.

## Replace the Pact Mock Server with an echo server

To confirm that it's not caused by `HttpClient` itself, I ran a local [Mock Server](https://www.mock-server.com/), like this:

```bash
docker run --rm -it -p 1080:1080 mockserver/mockserver
```

I set up the following hard-coded expectation (which is one of the responses expected by one of the example consumer tests) and hard-coded the same example consumer test to use the echo server on http://localhost:1080 instead of the Pact Mock Server:

```bash
curl -v -X PUT "http://localhost:1080/mockserver/expectation" -d '{
    "httpRequest": {
        "path": "/events"
    },
    "httpResponse": {
        "statusCode": 200,
        "headers": {
            "content-type": ["application/json"]
        },
        "body": [{
            "eventId": "45d80d13-d5a2-48d7-8353-cbb4c0eaabf5",
            "eventType": "SearchView",
            "timestamp": "2014-06-30T01:37:41.0660548"
        }, {
            "eventId": "83f9262f-28f1-4703-ab1a-8cfd9e8249c9",
            "eventType": "DetailsView",
            "timestamp": "2014-06-30T01:37:52.2618864"
        }, {
            "eventId": "3e83a96b-2a0c-49b1-9959-26df23f83aeb",
            "eventType": "SearchView",
            "timestamp": "2014-06-30T01:38:00.8518952"
        }]
    }
}'
```

The test executed in 131ms as I'd expect, with only 52ms of that being the HTTP request, so that proves that it's caused by the Pact Mock Server:

```
19:28:43.812 - Finished setup
19:28:43.824 - Starting verify
19:28:43.835 - Starting get events
19:28:43.836 -     Enter GetAllEvents
19:28:43.837 -     Sending request...
19:28:43.889 -     Performed request
19:28:43.889 -     Response is OK
19:28:43.890 -     Read content
19:28:43.914 -     Deserialised response
19:28:43.915 -     Started Dispose
19:28:43.915 -     Finished Dispose
19:28:43.915 - Starting assert
19:28:43.943 - Finished verify
```

**Comment by adamrodger** [comment]
I ran Fiddler so I could inspect the request and see whether it happens 'inside' the Pact mock server, and suddenly the performance is absolutely fine:

Fiddler:
![image](https://user-images.githubusercontent.com/508586/133939809-7c3ce581-5f47-4b0d-bae3-6a16987c6e94.png)

Test:
![image](https://user-images.githubusercontent.com/508586/133939823-1ef34812-79a0-4364-84c9-ae837a8fdf90.png)

Logs:
```
20:05:25.967 - Starting test
20:05:25.975 - Starting setup
20:05:26.064 - Finished setup
20:05:26.067 - Starting verify
20:05:26.072 - Starting get events
20:05:26.073 -     Enter GetAllEvents
20:05:26.073 -     Sending request...
20:05:26.122 -     Performed request
20:05:26.122 -     Response is OK
20:05:26.123 -     Read content
20:05:26.146 -     Deserialised response
20:05:26.146 -     Started Dispose
20:05:26.146 -     Finished Dispose
20:05:26.146 - Starting assert
20:05:26.175 - Finished verify
Mock server logs:

[DEBUG][pact_mock_server::hyper_server] Creating pact request from hyper request
[DEBUG][pact_mock_server::hyper_server] Extracting query from uri /events
[INFO][pact_mock_server::hyper_server] Received request Request ( method: GET, path: /events, query: None, headers: Some({"accept": ["application/json"], "host": ["127.0.0.1:56244"], "authorization": ["Bearer SomeValidAuthToken"]}), body: Empty )
[INFO][pact_matching] comparing to expected Request ( method: GET, path: /events, query: None, headers: Some({"Accept": ["application/json"], "Authorization": ["Bearer SomeValidAuthToken"]}), body: Missing )
[DEBUG][pact_matching]      body: ''
[DEBUG][pact_matching]      matching_rules: MatchingRules { rules: {PATH: MatchingRuleCategory { name: PATH, rules: {} }, HEADER: MatchingRuleCategory { name: HEADER, rules: {} }} }
[DEBUG][pact_matching]      generators: Generators { categories: {} }
[DEBUG][pact_matching::matchers] String -> String: comparing '/events' to '/events' using Equality
[DEBUG][pact_matching] expected content type = '*/*', actual content type = '*/*'
[DEBUG][pact_matching] content type header matcher = 'None'
[DEBUG][pact_matching::matchers] String -> String: comparing 'Bearer SomeValidAuthToken' to 'Bearer SomeValidAuthToken' using Equality
[DEBUG][pact_matching] --> Mismatches: []
[DEBUG][pact_mock_server::hyper_server] Test context = {"mockServer": Object({"href": String("http://127.0.0.1:56244"), "port": Number(56244)})}
[INFO][pact_mock_server::hyper_server] Request matched, sending response Response ( status: 200, headers: Some({"Content-Type": ["application/json; charset=utf-8"]}), body: Present(356 bytes) )
[DEBUG][pact_mock_server::hyper_server]      body: '[{"eventId":"45d80d13-d5a2-48d7-8353-cbb4c0eaabf5","eventType":"SearchView","timestamp":"2014-06-30T01:37:41.0660548"},{"eventId":"83f9262f-28f1-4703-ab1a-8cfd9e8249c9","eventType":"DetailsView","timestamp":"2014-06-30T01:37:52.2618864"},{"eventId":"3e83a96b-2a0c-49b1-9959-26df23f83aeb","eventType":"SearchView","timestamp":"2014-06-30T01:38:00.8518952"}]'

20:05:26.178 - Finished test
```

That proves it's fundamentally not the example code, which is good because that's just standard idiomatic .Net code. When either a proxy is in the way (such as Fidder) or you swap out the Pact mock server for a dumb echo server, the performance is absolutely fine. It's only when the Pact mock server is used that the performance problems occur.

**Comment by adamrodger** [comment]
OK, here's a really weird thing 😁 

If you use the standard approach (where the `baseUri` is `http://localhost:<port>` of the mock server) then the tests are really slow:

![image](https://user-images.githubusercontent.com/508586/133940161-7962e734-104e-4391-85f8-3fceafa06d17.png)
![image](https://user-images.githubusercontent.com/508586/133940175-aaf1dafc-e15a-4f8f-afa0-8008302a4f0e.png)

BUT if you replace `localhost` with `127.0.0.1` then the tests are fast:

![image](https://user-images.githubusercontent.com/508586/133940124-d9f50826-c071-4bb7-9d09-77f38db56e33.png)
![image](https://user-images.githubusercontent.com/508586/133940141-9cc25b27-4fc6-492d-94ab-2dbbdc30e3df.png)

I can't really explain why that happens at the moment? Perhaps the `hyper` server has some issue or different behaviour on Windows when you use `localhost` instead of `127.0.0.1`?

**Comment by adamrodger** [comment]
Erm, well this code looks rather problematic and I have no idea how it's even working:

![image](https://user-images.githubusercontent.com/508586/133940535-d5af27f2-f70f-4b42-9f36-074e4ef11839.png)

`this.host` is set to `IpAddress.Loopback`, which is just an enum defined in PactNet with nothing special about it at all. As far as I can tell that means the URL should (incorrectly) be set to `http://Loopback:<port>` but instead it gets set to `http://localhost:<port>`.

This is likely because the `IpAddress.Loopback.ToString() == "0"` and `http://0:<port>` is invalid, and thus `new Uri("http://0:<port>)` just defaults to `localhost` when it fails to parse 😱 

This is a typo in the original though - it should be using the IP address which is resolved further up so that the address resolves to `http://127.0.0.1:<port>` instead, which also works around the performance issue.

Edit: No, it genuinely does generate `http://Loopback:<port>` but then for some reason `new Uri()` converts that to `localhost`:

![image](https://user-images.githubusercontent.com/508586/133940685-f3bb87e7-4fdd-45a9-a2e4-632d06e60156.png)

I have no idea why it does that, because you could legitimately be trying to use a hostname of a machine called `Loopback`...?

**Comment by adamrodger** [comment]
From some googling, it appears as though Windows first tries to resolve `localhost` to `::1` (the IPV6 loopback), times out and then tries `127.0.0.1` (the IPv4 loopback). That would explain the strange performance problem using `localhost`.

I'm assuming the Rust FFI only listens on IPV4 or something, and this is why we experience that problem.

**Comment by mefellows** [comment]
Ah! Great find Adam 👏 .

**Comment by bethesque** [comment]
I had to laugh at "This is quite a large regression on the Ruby backend performance", because this is a sentence nobody has ever said before 😆

**Comment by adamrodger** [comment]
This issue can be closed now since the linked PR fixes it (referencing `127.0.0.1` explicitly instead of `localhost`), but I'll leave the related upstream issue open because the Pact mock server should probably also listen on the IPV6 loopback.
